<a href="https://colab.research.google.com/github/noviytrendASU/123123/blob/main/%D0%9B%D0%B0%D0%B1%D0%B04_%D0%9A%D0%B0%D1%88%D0%B8%D1%80%D0%BE%D0%B2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Анализ и обработка результатов

В анализе данных важно наглядное представление результатов, чтобы их мог оценить заказчик. Это бизнес, где на кону громадный трафик, серверные мощности, личное время многотысячной аудитории. И одновременно это исследование.

Как в настоящей экспериментальной науке, ответ непредсказуем. Его точность зависит от умелого владения статистическими методами и от качества исходных данных. Перед тем, как начинать считать, проверьте, грамотно ли ваши данные подготовлены.

# Группировка данных

Слово «анализ» означает разбор, рассмотрение с разных сторон. Анализ данных начинают с разделения их на группы по какому-нибудь признаку. Эта операция называется группировка данных. Она помогает изучить материал более подробно, чтобы затем перейти к поиску взаимосвязей между отдельными группами.


Группировка оправданна, если данные чётко делятся по значимому признаку, а полученные группы близки к теме задачи. Например, когда есть данные обо всех покупках в супермаркете, можно смело заниматься группировкой. Так можно установить время наплыва покупателей и решить проблему пиковых нагрузок. Или посчитать средний чек — обычно для магазинов это ключевая метрика.


Стадии группировки хорошо укладываются в словесную формулу **split-apply-combine**:

•	разделить, **split** — разбиение на группы по определённому критерию;

•	применить, **apply** — применение какого-либо метода к каждой группе в отдельности, например, подсчёт численности группы методом count() или суммирование вызовом sum();

•	объединить, **combine** — сведение результатов в новую структуру данных, в зависимости от условий разделения и выполнения метода это бывает DataFrame и Series.

На картинке изображен принцип **split-apply-combine** для таблицы с экзопланетами. Посмотрим, как вообще идут дела с поиском экзопланет. Сначала данные делят по группам, где каждая группа — это год. Потом метод **count()** подсчитывает численность каждой группы. В итоге получаем новую структуру данных с группами, где каждая содержит год и число открытых за этот год экзопланет.

![image.png](attachment:image.png)

В Рandas для группировки данных есть метод **groupby()**. Он принимает как аргумент название столбца, по которому нужно группировать. В случае с делением экзопланет по годам открытия:

**print(exoplanet.groupby('discovered'))**

**<pandas.core.groupby.DataFrameGroupBy object at 0x7fc1e1ca3400>**


Применение метода **groupby()** к объекту типа DataFrame приводит к созданию объекта особого типа — **DataFrameGroupBy**. Это сгруппированные данные. Если применить к ним какой-нибудь метод Pandas, они станут новой структурой данных типа **DataFrame** или **Series**.
Подсчитаем сгруппированные по годам экзопланеты методом **count()**:

Если нужно сравнить наблюдения по одному показателю, метод применяют к **DataFrameGroupBy** с указанием на один столбец. Нас в первую очередь интересует радиус экзопланет: мы ищем другую Землю. Давайте получим таблицу с единственным столбцом 'radius':

**exo_number = exoplanet.groupby('discovered')['radius'].count()**

**print(exo_number)**

Получили Series, где по годам открытия расписано количество экзопланет, для которых удалось установить радиус.

Посмотрим, как меняется средний радиус открытых экзопланет год от года. Для этого надо сложить радиусы планет, открытых за определённый год, и поделить на их количество (которое мы уже нашли).

Сумма радиусов считается методом **sum()**:

**exo_radius_sum = exoplanet.groupby('discovered')['radius'].sum()**


**print(exo_radius_sum)**



## Задача
Вашему вниманию предлагается датасет с обзором показателей самоубийств с 1985 по 2016 год — Сравнение социально-экономической информации с показателями самоубийств по годам и странам.

Столбцами этой таблицы являются: страна, год, пол, возрастная группа, количество самоубийств, население, уровень самоубийств, сводный ключ по стране за год, ИЧР (индекс человеческого развития) за год, ВВП за год, ВВП на душу населения, поколение (на основе среднего по возрастной группе).

1. Загрузите таблицу master, содержащую данные о самоубийствах. Выведите на экран таблицу и оцените данные, обратите внимание на названия столбцов, переименуйте их, если это необходимо.

2. Убедитесь в том, что данные прошли предподготовку.  Пропущенные и неопределённые значения выявляет метод isna(), а суммарное количество таких значений — метод sum().

3. Сгруппируйте DataFrame по столбцу country, сохраните полученный результат в переменной country_grouping.

4. Посчитайте количество случаев самоубийств по странам, которые были совершены, методом **sum()**, указав, что выбираем один столбец suicides_no.

Сохраните результат в переменной country_sum и выведите первые 30 строк этой таблицы.



In [2]:
import pandas as pd
master=pd.read_csv('master.csv')
master = master.set_axis(['country', 'year', 'sex', 'age', 'suicides_no', 'population', 'suicides_per_100k_pop', 'country_year', 'hdi_for_year', 'gdp_for_year', 'gdp_per_capita', 'generation'], axis = 'columns')
print('Первые 10 строк таблицы с обновленными названиями:\n', master.head(10))
print(master.info())
print(master.describe())

Первые 10 строк таблицы с обновленными названиями:
    country  year     sex          age  suicides_no  population  \
0  Albania  1987    male  15-24 years           21      312900   
1  Albania  1987    male  35-54 years           16      308000   
2  Albania  1987  female  15-24 years           14      289700   
3  Albania  1987    male    75+ years            1       21800   
4  Albania  1987    male  25-34 years            9      274300   
5  Albania  1987  female    75+ years            1       35600   
6  Albania  1987  female  35-54 years            6      278800   
7  Albania  1987  female  25-34 years            4      257200   
8  Albania  1987    male  55-74 years            1      137500   
9  Albania  1987  female   5-14 years            0      311000   

   suicides_per_100k_pop country_year  hdi_for_year   gdp_for_year  \
0                   6.71  Albania1987           NaN  2,156,624,900   
1                   5.19  Albania1987           NaN  2,156,624,900   
2          

In [3]:
print("\nКоличество пропущенных и неопределенных значений:\n",master.isna().sum())
master['hdi_for_year'] = master['hdi_for_year'].fillna('unknown')
print("\nОбновленное количество пропущенных и неопределенных значений:\n",master.isna().sum())


Количество пропущенных и неопределенных значений:
 country                      0
year                         0
sex                          0
age                          0
suicides_no                  0
population                   0
suicides_per_100k_pop        0
country_year                 0
hdi_for_year             19456
gdp_for_year                 0
gdp_per_capita               0
generation                   0
dtype: int64

Обновленное количество пропущенных и неопределенных значений:
 country                  0
year                     0
sex                      0
age                      0
suicides_no              0
population               0
suicides_per_100k_pop    0
country_year             0
hdi_for_year             0
gdp_for_year             0
gdp_per_capita           0
generation               0
dtype: int64


In [4]:
country_grouping = master.groupby('country')
print(country_grouping.describe)

<bound method GroupBy.describe of <pandas.core.groupby.generic.DataFrameGroupBy object at 0x7d5b37d09210>>


In [5]:
country_sum = country_grouping['suicides_no'].sum().reset_index()
print('\nКоличество самоубийств по странам:\n', country_sum.head(20))


Количество самоубийств по странам:
                    country  suicides_no
0                  Albania         1970
1      Antigua and Barbuda           11
2                Argentina        82219
3                  Armenia         1905
4                    Aruba          101
5                Australia        70111
6                  Austria        50073
7               Azerbaijan         1656
8                  Bahamas           93
9                  Bahrain          463
10                Barbados          177
11                 Belarus        59892
12                 Belgium        62761
13                  Belize          348
14  Bosnia and Herzegovina          318
15                  Brazil       226613
16                Bulgaria        36388
17              Cabo Verde           42
18                  Canada       107561
19                   Chile        40895


# Сортировка данных


Поиск необычного в группе — что среди планет, что среди меломанов — это прежде всего поиск чемпионов: объектов с выдающимися показателями по разным статьям. Как всю таблицу, так и отдельные группы изучают, сортируя строки по какому-либо столбцу.

В Pandas для этой операции есть метод **sort_values()**. У него два аргумента:

• **by = 'имя столбца'** — имя столбца, по которому нужно сортировать;

• **ascending:** по умолчанию True. Для сортировки по убыванию установите значение False.

![image.png](attachment:image.png)


Среди экзопланет интересны близкие по размерам к Земле. Есть ли такие? Отсортируем список по радиусу в порядке возрастания. Тогда в голове таблицы окажутся самые малые, на которых гравитация не прижмёт нас к полу.

**print(exoplanet.sort_values(by='radius').head(30))**

Оказывается, некоторые из уже открытых экзопланет по размерам близки не то что к Земле, но уже и к Луне! Получим список экзопланет с радиусом меньше земного. Смотрите, как логический оператор (здесь это <) задействуется в отборе элементов столбца. Дальше нам этот приём не раз понадобится.

**print(exoplanet[exoplanet['radius'] < 1])**

Но и этот список такой длинный, что изучать его лучше по частям. Экзопланеты, близкие по размерам к Земле, за последнее десятилетие открывали нередко. Можно изучать список открытых за каждый год. Например, для 2014 года (вновь обратите внимание на работу логического оператора, теперь это ==):

**print(exoplanet[exoplanet['discovered'] == 2014])**

А чтобы не тратить время на лишнее, поставим оба условия сразу. Для этого в Pandas есть логический оператор &, подобный оператору and языка Python. Напомним, его смысл на русском языке можно передать словами «и ещё»:

**exo_small_14 = exoplanet[ (exoplanet['radius']<1) & (exoplanet['discovered']==2014)]**

**print(exo_small_14)**

Отсортируем результат в порядке убывания радиуса.

**print(exo_small_14.sort_values(by = 'radius', ascending = False))**





## Задача
1. Выделим пятерки стран с самым низким и самым высоким уровнем самоубийств в мире.

In [6]:
country_suicides = master.groupby('country')['suicides_no'].sum().reset_index()
lowest_country_suicides = country_suicides.sort_values(by='suicides_no', ascending=True).head(5)
biggest_country_suicides = country_suicides.sort_values(by='suicides_no', ascending=False).head(5)

print('Пятерка стран с самым низким уровнем самоубийств:\n', lowest_country_suicides)
print('\nПятерка стран с самым высоким уровнем самоубийств:\n', biggest_country_suicides)

Пятерка стран с самым низким уровнем самоубийств:
                   country  suicides_no
27               Dominica            0
76  Saint Kitts and Nevis            0
79             San Marino            4
1     Antigua and Barbuda           11
55               Maldives           20

Пятерка стран с самым высоким уровнем самоубийств:
                country  suicides_no
75  Russian Federation      1209742
98       United States      1034013
46               Japan       806902
33              France       329127
95             Ukraine       319950


2. Исследуйте распределение количества суидидов по гендерному признаку. Найдите лидеров и антилидеров по количеству самоубийств среди женщин. Повторяется ли картина по сравнению с предыдущим исследованием?

In [7]:
women_suicides = master[master['sex'] == 'female']
women_grouping = women_suicides.groupby('country')['suicides_no'].sum().reset_index()

leaders_suicides = women_grouping.sort_values(by = 'suicides_no', ascending = False).head(10)
anti_leaders_suicides = women_grouping.sort_values(by = 'suicides_no', ascending = True).head(10)

print('Лидеры по самоубийствам среди женщин:\n', leaders_suicides)
print('Антилидеры по самоубийствам среди женщин:\n', anti_leaders_suicides)

Лидеры по самоубийствам среди женщин:
                country  suicides_no
46               Japan       251630
75  Russian Federation       214330
98       United States       213797
33              France        89419
73   Republic of Korea        82615
35             Germany        79707
95             Ukraine        61377
15              Brazil        49015
44               Italy        32881
97      United Kingdom        32687
Антилидеры по самоубийствам среди женщин:
                   country  suicides_no
65                   Oman            0
79             San Marino            0
27               Dominica            0
76  Saint Kitts and Nevis            0
1     Antigua and Barbuda            1
37                Grenada            4
17             Cabo Verde            5
55               Maldives            6
81             Seychelles            9
48               Kiribati           10


# Описательная статистика

Чтобы представить себе среднестатистического самоубийцу, понадобится описательная статистика.

Из её показателей для количественного описания данных в нашей задаче нужны четыре меры — максимум, минимум, медиана и среднее.

Наибольшее и наименьшее обычно вычисляют только по одному признаку. Например, можно получить минимальное и максимальное значение уровня самоубийств (количество самоубийств на 100 тыс.населения) (столбец 'suicides/100k pop'). Для поиска максимума вызывают метод max(), примененный к этому столбцу:

print(master['suicides/100k pop'].max())


In [8]:
print(master['suicides_per_100k_pop'].max())

224.97


Самый высокий показатель оказался 224.97 самоубийств на 100 тыс.населения. Интересно, в какой стране такая ситуация.

Запросим из master строку с максимальным значением, прибегнув к логической индексации с условием

**master['suicides/100k pop'] == master['suicides/100k pop'].max()**

In [9]:
master['suicides_per_100k_pop'] == master['suicides_per_100k_pop'].max()

0        False
1        False
2        False
3        False
4        False
         ...  
27815    False
27816    False
27817    False
27818    False
27819    False
Name: suicides_per_100k_pop, Length: 27820, dtype: bool

Минимальное значение — ищут методом **min()**. Нам интересно установить случаи, когда уровень суицидов минимальный, но не нулевой.

Вот как мы создаём выборку  и находим в ней минимальное значение:

**master_drop_null = master[master['suicides/100k pop'] != 0]**

**print(master_drop_null['suicides/100k pop'].min())**


In [10]:
master_drop_null = master[master['suicides_per_100k_pop'] != 0]
print(master_drop_null['suicides_per_100k_pop'].min())

0.02


Получим название самой "счастливой" страны. Как и при поиске максимума, воспользуемся условием и логической индексацией

In [11]:
min_index = master_drop_null['suicides_per_100k_pop'].idxmin()
min_suicides_country = master_drop_null.loc[min_index, 'country']
print('Самая счастливая страна:', min_suicides_country)

Самая счастливая страна: South Africa


Результат вывода — две страны, которые соответствуют минимальному значению.
На основе полученных данных можно сделать вывод, что число совершенных самоубийств на 100 тыс. населения  находится в диапазоне от 0,02 до 224.97 секунд, не включая пропущенные.

Это знание пригодится, чтобы разобраться с медианой и средним арифметическим.

Среднее и медиана оценивают значения в центре выборки. Если тех строк, где число с большим количеством самоубийц, столько же, сколько тех, где это число мало — среднее подойдет. Но когда есть оторванные от основной массы лидеры, их результаты сильно смещают значение среднего вверх. Вот почему оценивать в этом случае лучше медианой.

![image.png](attachment:image.png)

В геометрии медиана делит фигуру на две равные по площади. В статистике она делит выборку пополам: в одной половине значения меньше медианного, в другой больше. Логично, что для определения медианы список обязательно должен быть отсортирован — либо по возрастанию, либо по убыванию.
Когда количество значений нечётное, медиана будет равна тому значению, которое оказалось ровно посередине отсортированного набора. Если же количество данных чётное, то медиана рассчитывается как среднее арифметическое двух соседних чисел в середине набора.



В Pandas есть метод **median()**, который считает медиану. По аналогии с **min()** и **max()** его можно применять ко всей таблице, к отдельному столбцу или к сгруппированным данным. Теперь применим медиану ко всем значениям времени прослушивания в нашей таблице, исключив нулевые:



In [12]:
print('Медиана:', master_drop_null['suicides_per_100k_pop'].median())

Медиана: 8.26


Чтобы убедиться, что лидеры действительно смещают средний показатель вверх, найдём среднее арифметическое всех этих значений методом mean():

In [13]:
print('Среднее:', master[master['suicides_per_100k_pop'] != 0]['suicides_per_100k_pop'].mean())

Среднее: 15.14694039678831
